File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/operation/Shape2DBinaryOperation.java
package com.greentree.commons.geometry.geom2d.operation;

import com.greentree.commons.geometry.geom2d.IShape2D;
import com.greentree.commons.geometry.geom2d.Shape2DUtil;
import com.greentree.commons.geometry.geom2d.collision.CollisionEvent2D;
import com.greentree.commons.math.MathLine2D;
import com.greentree.commons.math.vector.AbstractVector2f;
import com.greentree.commons.math.vector.Vector2f;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

public class Shape2DBinaryOperation<A extends IShape2D, B extends IShape2D> {

    public static final Shape2DBinaryOperation DEFAULT = new Shape2DBinaryOperation();

    public CollisionEvent2D.Builder getCollisionEvent(A a, B b) {
        var normal = Shape2DUtil.getCollisionNormalOnNormalProjection(a, b);
        return new CollisionEvent2D.Builder(Shape2DUtil.getCollisionPoint(a, b), normal,
                Shape2DUtil.getProjectionOverlay(a, b, new Vector2f(normal.y(), -normal.x())));
    }

    public Collection<AbstractVector2f> getContactPoint(final A a, final B b) {
        final Set<AbstractVector2f> res = new HashSet<>();
        for (final var al : a.getLinesLoop())
            for (final var bl : b.getLinesLoop()) {
                final Vector2f c = MathLine2D.contact(al.getMathLine(), bl.getMathLine());
                if (c == null || !al.getAABB().isInside(c.x(), c.y()) || !bl.getAABB().isInside(c.x(), c.y()))
                    continue;
                res.add(c);
            }
        return res;
    }

    public boolean isIntersect(final A a, final B b) {
        if (!a.getAABB().isIntersect(b.getAABB()))
            return false;
        return Shape2DUtil.isIntersectOnNormalProjection(a, b);
    }

}

File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/operation/Shape2DBinaryOperation.kt
package com.greentree.commons.geometry.geom2d.operation

import com.greentree.commons.geometry.geom2d.Shape2D
import com.greentree.commons.geometry.geom2d.Shape2DUtil
import com.greentree.commons.geometry.geom2d.collision.CollisionEvent2D
import com.greentree.commons.geometry.geom2d.isIntersect
import com.greentree.commons.geometry.geom2d.shape.FiniteShape2D
import com.greentree.commons.math.MathLine2D
import org.joml.Vector2f
import org.joml.Vector2fc

open class Shape2DBinaryOperation<A : Shape2D, B : Shape2D> {

	open fun getCollisionEvent(a: A, b: B): CollisionEvent2D.Builder {
		a as FiniteShape2D
		b as FiniteShape2D
		val normal = Shape2DUtil.getCollisionNormalOnNormalProjection(a, b)
		return CollisionEvent2D.Builder(
			Shape2DUtil.getCollisionPoint(a, b), normal,
			Shape2DUtil.getProjectionOverlay(a, b, Vector2f(normal.y(), -normal.x()))
		)
	}

	open fun getContactPoint(a: A, b: B): Collection<Vector2fc> {
		a as FiniteShape2D
		b as FiniteShape2D
		val res: MutableSet<Vector2fc> = HashSet()
		for(al in a.linesLoop)
			for(bl in b.linesLoop) {
				val c = MathLine2D.contact(al.mathLine, bl.mathLine)
				if(c == null || !al.boundingBox.isInside(c.x(), c.y()) || !bl.boundingBox.isInside(
						c.x(),
						c.y()
					)
				) continue
				res.add(c)
			}
		return res
	}

	open fun isIntersect(a: A, b: B): Boolean {
		a as FiniteShape2D
		b as FiniteShape2D
		if(!a.boundingBox.isIntersect(b.boundingBox))
			return false
		return Shape2DUtil.isIntersectOnNormalProjection(a, b)
	}

	companion object {

		@JvmField
		val DEFAULT = Shape2DBinaryOperation<Shape2D, Shape2D>()
	}
}
